### PR TITLE
Interface with the distributed MUMPS solver

### PR DESCRIPTION
**Pull request by psocratis**
This PR adds an interface to MUMPSSolver. 

Notes for the reviewers:

1. There are two implementations with regards to how the RHS and the Solution vectors are handled. Upto MUMPS version 
    5.3.0 there was no support for distributed RHS vectors and a global RHS vector had to be assembled on host. In this PR 
    both cases are supported and for versions >=5.3.0 the distributed RHS and Solution case is preferred. 

2. In the case of distributed solution vectors, MUMPS does not necessarily return them in the same parallel partition as the 
    RHS are given. Therefore the solution vectors need to be redistributed.  

3.  ex25p was modified to demonstrate the interface (mimicking ex11p). 

TODOs

- [x] Distributed matrix case
- [x] RHS and Sol on host 
- [x] distributed RHS
- [x] distributed Sol
- [x] Adding support for Sym and SPD matrices
- [x] Adding MultTranspose
- [x] Adding example run in ex25p
- [x] Passing valgrind checks

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1788](https://github.com/mfem/mfem/pull/1788) | @psocratis | @v-dobrev | @vladotomov + @jakubcerveny | 10/06/20 | 10/22/20 | 11/03/20 | |

### CODE REVIEW

File: INSTALL
Comment (vladotomov) [inline INSTALL:577]: You should mention that it also requires `MFEM_USE_LAPACK=YES`, `MFEM_USE_SCALAPACK=YES`.

File: linalg/mumps.hpp
Comment (vladotomov) [inline linalg/mumps.hpp:75]: You should state that the `Set*` methods must be called before `SetOperator` to have any effect.

**Review by vladotomov — APPROVED** [review]
The PR is well organized and prepared diligently.

I haven't run the code -- I tried building the MUMPS and all its dependancies on my mac, but hit some issues and don't have the time to continue. Sorry about that.

If the reviewers are expected to build and test in such cases, then the authors should be expected to provide a sample `user.mk` file, and maybe some comments about the external package build.

File: INSTALL
Comment (v-dobrev) [reply INSTALL:577]: If LAPACK and ScaLAPACK are needed for MUMPS that does not necessarily mean you need to enable them in MFEM too (by setting `MFEM_USE_LAPACK=YES` and/or `MFEM_USE_SCALAPACK=YES`). What you can do is just use `LAPACK_OPT`, `LAPACK_LIB`, `SCALAPACK_OPT`, and `SCALAPACK_LIB` when defining `MUMPS_OPT` and `MUMPS_LIB`. In other words, the options `MFEM_USE_LAPACK` and `MFEM_USE_MUMPS` should be independent.

Also, I don't think the option `MFEM_USE_SCALAPACK` really does anything, so it should probably be removed. Of course, the variables `SCALAPACK_OPT` and `SCALAPACK_LIB` should stay since they will allow proper configuration of the `MUMPS_*` variables, as mentioned above.

Comment (psocratis) [reply INSTALL:577]: There is no MFEM_USE_SCALAPACK option.

File: linalg/mumps.hpp
Comment (psocratis) [reply linalg/mumps.hpp:75]: Good point!

**Comment by psocratis** [comment]
> The PR is well organized and prepared diligently.
> 
> I haven't run the code -- I tried building the MUMPS and all its dependancies on my mac, but hit some issues and don't have the time to continue. Sorry about that.
> 
> If the reviewers are expected to build and test in such cases, then the authors should be expected to provide a sample `user.mk` file, and maybe some comments about the external package build.

Hi @vladotomov, that's a good point. I am attacking my user.mk file for mfem and the Makefile.inc file for MUMPS just in case you want to try it out. 

[files.zip](https://github.com/mfem/mfem/files/5398292/files.zip)

**Comment by jakubcerveny** [comment]
Do I need ParMetis to run this? Can I use (binary package) MUMPS 5.2.1 straight from Ubuntu 20.04? A little info how to set this up would be helpful...

File: linalg/mumps.cpp
Comment (jakubcerveny) [inline linalg/mumps.cpp:43]: These two lines are failing for me since I have hypre built with `--bigint` and therefore `HYPRE_Int` == `long long int*`. If MUMPS supports 64-bit matrix indices, we should also support that, otherwise a compile time error should be printed that there is a conflicting configuration of hypre.

**Comment by jakubcerveny** [comment]
OK so I managed to set this up with `libmetis-dev`, `libparmetis-dev`, `libscalapack-openmpi-dev`, `libmumps-dev` from Ubuntu 20.04, and MFEM_USE_METIS_5. I also had to remove `-lmpifort` and change `-lscalapack` to `-lscalapack-openmpi`. I'll leave it to the editor if the makefile can be improved or not (I'm not an expert and maybe there's a better way to set this up).

`ex25p` seems to be running much faster with `-mumps`. Very nice example BTW. Still looking...

**Review by jakubcerveny — CHANGES_REQUESTED** [review]
Apart from the 64-bit indices, this looks good!

**Comment by jakubcerveny** [comment]
(MUMPS seems to be extremely fast for 2D problems, but in 3D it seems to get into trouble very quickly, I guess that's another story though)

File: linalg/mumps.cpp
Comment (psocratis) [reply linalg/mumps.cpp:43]: good point! I'll fix it

Comment (psocratis) [reply linalg/mumps.cpp:47]: MUMPS by default supports 64-bit ints only for the `nnz_loc` counter.  In case we want full support for 64-bit ints, MUMPS should be compiled with C preprocessing flag `-DINTSIZE64` and Fortran compiler option `-i8` (or equivalent). Additionally, all the required libraries have to be compiled with 64-bit support (MPI, BLACS,ScaLAPACK, LAPACK , BLAS etc). I am not sure that this is crucial, at least for now, so I decided to add a check for Hypre BIGINT (same check we do for SuperLUdist), which will print a compile time error. Sounds good?

**Review by jakubcerveny — APPROVED** [review]

File: INSTALL
Comment (v-dobrev) [inline INSTALL:590]: `CHANGELOG` says `versions >= 5.2.0` -- which one is correct?

Comment (psocratis) [reply INSTALL:590]: it is 5.1.1. I fixed it.

File: linalg/mumps.hpp
Comment (v-dobrev) [inline linalg/mumps.hpp:120]: This is not meant to be used by users, right? If that's the case, it is better to move these macros to the `.cpp` file.

Comment (psocratis) [reply linalg/mumps.hpp:121]: correct

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Merged in post v4.2 `next`.

**Comment by v-dobrev** [comment]
Reminder: move `CHANGELOG` entry for this branch to the new v4.2.1 section after #1857 is merged.

**Comment by tzanio** [comment]
Addressed https://github.com/mfem/mfem/pull/1788#issuecomment-720075595

Re-merged in `next` for testing ...

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Pull request merged**
